src/transformStateWithClones.js

```diff
@@ -6,8 +6,55 @@
  *
  * @return {Object[]}
  */
+/* eslint no-console: ["error", { allow: ["warn", "error", "log"] }] */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+
+  for (const item of actions) {
+    if (item.type === 'addProperties') {
+      // if our 'result' arr has no therefore we use state obj
+      if (result.length === 0) {
+        const objWithAddedProps = Object.assign({}, state, item.extraData);
+
+        result.push(objWithAddedProps);
+      } else {
+        // if our 'result' arr has last item then we use last item
+        const objWithAddedProps = Object.assign(
+          {},
+          result[result.length - 1],
+          item.extraData,
+        );
+
+        result.push(objWithAddedProps);
+      }
+    }
+
+    if (item.type === 'removeProperties') {
+      let objWithRemovedProps;
+      // if our 'result' arr has no therefore we use state obj
+
+      if (result.length === 0) {
+        objWithRemovedProps = Object.assign({}, state);
+
+        // if our 'result' arr has last item then we use last item
+      } else {
+        objWithRemovedProps = Object.assign({}, result[result.length - 1]);
+      }
+
+      for (const key in item.keysToRemove) {
+        const propKeyToRemove = item.keysToRemove[key];
+
+        delete objWithRemovedProps[propKeyToRemove];
```

Comment:
[NAMING] - The variable 'objWithRemovedProps' could be named better to reflect that it is a copy of the state. Consider renaming it to 'stateCopyWithRemovedProps' or something similar for clarity.